### PR TITLE
[Gateway] Implement StakeGateway Message and Add Tests

### PR DESCRIPTION
**Pull request by h5law**
### Human Summary

- Add stake field to the stake and gateway proto messages
- Update validation
- Update and add unit tests relating to
  - genesis
  - cli
  - staking
- Hydrate the stake keeper method with the correct logic
- Add makefile helpers

## Issue

Fixes (in part) [19](https://github.com/pokt-network/poktroll/issues/19)

## Type of change

Select one or more:

- [x] New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing

- [x] **Run all unit tests**: `make test_all_unit`
- [x] **Verify Localnet manually**: See the instructions [here](TODO: add link to instructions)

## Sanity Checklist

- [x] I have tested my changes using the available tooling
- [x] I have performed a self-review of my own code
- [x] I have commented my code, updated documentation and left TODOs throughout the codebase

### CODE REVIEW

File: x/gateway/client/cli/tx_stake_gateway.go [hunk 1]
```diff
@@ -3,29 +3,35 @@ package cli
...
 	"github.com/spf13/cobra"
-	"pocket/x/gateway/types"
 )
 
 var _ = strconv.Itoa(0)
 
 func CmdStakeGateway() *cobra.Command {
 	cmd := &cobra.Command{
-		Use:   "stake-gateway",
+		Use:   "stake-gateway [amount]",
```
Comment (Olshansk) [inline x/gateway/client/cli/tx_stake_gateway.go:19]: See #59 for additions I made to usage & such

File: x/gateway/client/cli/tx_stake_gateway.go [hunk 2]
```diff
@@ -3,29 +3,35 @@ package cli
...
-		Args:  cobra.ExactArgs(0),
+		Args:  cobra.ExactArgs(1),
 		RunE: func(cmd *cobra.Command, args []string) (err error) {
-
 			clientCtx, err := client.GetClientTxContext(cmd)
 			if err != nil {
 				return err
 			}
-
+			stakeAmountString := args[0]
```
Comment (Olshansk) [inline x/gateway/client/cli/tx_stake_gateway.go:27]: s/stakeAmountString/stakeString
s/stakeAmount/stake

It includes both amount & denom

File: x/gateway/client/cli/tx_stake_gateway_test.go [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+		},
+		{
+			desc:        "stake gateway: valid",
+			address:     gatewayAccount.Address.String(),
+			stakeAmount: "1000upokt",
+		},
+	}
+
+	// Initialize the Gateway Account by sending it some funds from the validator account that is part of genesis
+	sendArgs := []string{
```
Comment (Olshansk) [inline x/gateway/client/cli/tx_stake_gateway_test.go:90]: See #59 for changes I made to leverage a helper.

File: x/gateway/keeper/msg_server_stake_gateway.go [hunk 4]
```diff
@@ -3,15 +3,81 @@ package keeper
...
+	msg *types.MsgStakeGateway,
+) error {
+	// Checks if the the msg address is the same as the current owner
+	if msg.Address != gateway.Address {
+		return errors.Wrapf(types.ErrGatewayUnauthorized, "msg Address (%s) != gateway address (%s)", msg.Address, gateway.Address)
+	}
+	if msg.Stake == nil {
+		return errors.Wrapf(types.ErrGatewayInvalidStake, "stake amount cannot be nil")
+	}
+	if gateway.Stake.IsGTE(*msg.Stake) {
```
Comment (Olshansk) [inline x/gateway/keeper/msg_server_stake_gateway.go:78]: @bryanchriswhite suggested use IsLTE here instead for readability - I agree.

File: x/gateway/types/errors.go [hunk 5]
```diff
@@ -1,12 +1,13 @@
...
 
-import (
-	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
-)
+// DONTCOVER
 
 // x/gateway module sentinel errors
 var (
-	ErrSample = sdkerrors.Register(ModuleName, 1100, "sample error")
+	ErrSample                = errors.Register(ModuleName, 1100, "sample error")
```
Comment (Olshansk) [inline x/gateway/types/errors.go:9]: Should we remove the sample?

File: x/gateway/types/genesis.go [hunk 6]
```diff
@@ -19,15 +21,32 @@ func DefaultGenesis() *GenesisState {
...
 			return fmt.Errorf("duplicated index for gateway")
 		}
 		gatewayIndexMap[index] = struct{}{}
+		// Validate the stake of each gateway
+		if elem.Stake == nil {
+			return fmt.Errorf("nil stake amount for gateway")
+		}
+		stakeAmount, err := sdk.ParseCoinNormalized(elem.Stake.String())
+		if !stakeAmount.IsValid() {
+			return fmt.Errorf("invalid stake amount for gateway %v; (%v)", elem.Stake, stakeAmount.Validate())
```
Comment (Olshansk) [inline x/gateway/types/genesis.go:39]: See #59 and return proper errors.

**Review by Olshansk — CHANGES_REQUESTED** [review]
Overall LGTM, but I think the big piece is just catching up with the changes I made in the other PR.

**Comment by Olshansk** [comment]
@h5law Also note that this is not "Hydration," but this is an implementation of the StakeGateway message. I recommend asking ChatGPT about the difference if you want to learn more.

Screenshot 2023-10-17 at 5 03 20 PM

**Comment by h5law** [comment]
@Olshansk have addressed the comments and caught up with #59 only thing I removed one of your helpers in the network test utils, I couldn't get it to work with the types. It only worked for the application genesis state and I couldn't figure out a good way to make it generic without stepping into encoding and decoding `any` types so thought best this be put into the helpers that call this method so simplicities sake.

File: cmd/pocketd/cmd/root.go [hunk 7]
```diff
@@ -39,8 +39,6 @@ import (
 	"github.com/spf13/cobra"
 	"github.com/spf13/pflag"
 
-	// this line is used by starport scaffolding # root/moduleImport
```
Comment (Olshansk) [inline cmd/pocketd/cmd/root.go]: Please add this back

File: x/application/client/cli/helpers_test.go [hunk 8]
```diff
@@ -24,6 +26,8 @@ func networkWithApplicationObjects(t *testing.T, n int) (*network.Network, []typ
 	t.Helper()
 	cfg := network.DefaultConfig()
 	appGenesisState := network.DefaultApplicationModuleGenesisState(t, n)
-	network.HydrateGenesisState(t, &cfg, appGenesisState, types.ModuleName)
+	buf, err := cfg.Codec.MarshalJSON(appGenesisState)
```
Comment (Olshansk) [inline x/application/client/cli/helpers_test.go:29]: If we're not going to use `HydrateGenesisState`, let's remove that funciton.

**Review by Olshansk — APPROVED** [review]
LGTM but PTAL at the two tiny comments before merging.

File: x/application/client/cli/helpers_test.go (hunk 8 above)
Comment (h5law) [reply x/application/client/cli/helpers_test.go:29]: Pretty sure I did

**Pull request merged**